File: MatlabCode/02/InventoryControl.m
%% File name: InventoryControl.m

%% Clean workspace

close all;
clear all;
clc;



%% Parameters

% Final simulation time
T_max = 100;

% Time horizon
T=[0:T_max-1]'; 



%% Initialization

% Inizialize matrices to store results of simulation runs
% Not necessary, but improves efficiency
x=zeros(T_max+1,1);
u=zeros(T_max,1);
gt=zeros(T_max,1);
w=zeros(T_max,1);



%% Single run simulation

% Simulate IC over t=0,...,T_max
% Dynamic system representation
[x, u, gt, w] = IC_singlerun(T_max);



%% Write results

% Write some results
disp(['        Time horizon T:  ' num2str(T_max) ]);
disp(' ');
disp(['            Total cost: ' num2str(sum(gt))]);
disp(['    First reorder time:    ' num2str(find(u,1)-1)]);
disp(['Total number of orders:   ' num2str(sum(u>0))]);



%% Figure 1: state

% Plot x vs. time
figure
hold on;
stairs([T; T_max],x,'LineWidth',2);
xlabel('t');
ylabel('x_t');
title('State (stock level)');
xlim([0 T_max])
grid;



%% Figure 2: input

% Plot u vs. time
figure
hold on;
bar(T,u);
xlabel('t');
ylabel('u_t');
title('Input (ordered quantity)');
xlim([0 T_max])
grid;



%% Figure 3: stage cost

% Plot gt vs. time
figure
hold on;
bar(T,gt);
xlabel('t');
ylabel('g_t');
title('Stage cost');
xlim([0 T_max])
grid;



%% Figure 4: total cost

% Plot J vs. time
figure
hold on;
area(T,cumsum(gt),'FaceColor','r');
xlabel('t');
ylabel('J');
title('Total cost');
xlim([0 T_max])
grid;

File: PythonCode/02/InventoryControl.py
import numpy as np
from IC_singlerun import IC_singlerun
import matplotlib.pyplot as plt

T_max = 100                                     # time horizon

x, u, gt, W_real = IC_singlerun(T_max)   # runs a single simulation over time horizon

print(f'Time horizon T: {T_max}')
print()
print(f'Total cost : {int(np.sum(gt))}')
print(f'First reorder time: {next((i for i, x in enumerate(u[0]) if x), None)}')
print(f'Total number of orders: {int(np.sum(u))}')

# plot
fig, axs = plt.subplots(2, 2)
fig.suptitle('Simulation')

axs[0, 0].stairs(x[:, 0], range(T_max + 2))
axs[0, 0].set_title('State (stock level)')

axs[0, 1].bar(range(T_max), u[:, 0])
axs[0, 1].set_title('Input (ordered quantity)')

axs[1, 0].bar(range(T_max), gt[:, 0])
axs[1, 0].set_title('Stage cost')

axs[1, 1].fill_between(range(T_max), np.cumsum(gt[:, 0]))
axs[1, 1].set_title('Total cost')

plt.show()
